java/ql/src/semmle/code/java/frameworks/JavaxJson.qll

```diff
@@ -8,55 +8,103 @@ private import semmle.code.java.dataflow.ExternalFlow
-        "javax.json;JsonNumber;false;doubleValue;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonNumber;false;intValue;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonNumber;false;intValueExact;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonNumber;false;longValue;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonNumber;false;longValueExact;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonObject;false;getBoolean;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonObject;false;getBoolean;;;Argument[1];ReturnValue;value",
-        "javax.json;JsonObject;false;getInt;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonObject;false;getInt;;;Argument[1];ReturnValue;value",
-        "javax.json;JsonObject;false;getJsonArray;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonObject;false;getJsonNumber;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonObject;false;getJsonObject;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonObject;false;getJsonString;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonObject;false;getString;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonObject;false;getString;;;Argument[1];ReturnValue;value",
-        "javax.json;JsonObjectBuilder;false;add;;;Argument[-1];ReturnValue;value",
-        "javax.json;JsonObjectBuilder;false;add;;;Argument[1];Argument[-1];taint",
-        "javax.json;JsonObjectBuilder;false;addNull;;;Argument[-1];ReturnValue;value",
-        "javax.json;JsonObjectBuilder;false;build;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonReader;false;read;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonReader;false;readArray;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonReader;false;readObject;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonReaderFactory;false;createReader;;;Argument[0];ReturnValue;taint",
-        "javax.json;JsonString;false;getChars;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonString;false;getString;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonValue;false;toString;;;Argument[-1];ReturnValue;taint",
-        "javax.json;JsonWriter;false;write;;;Argument[0];Argument[-1];taint",
-        "javax.json;JsonWriter;false;writeArray;;;Argument[0];Argument[-1];taint",
-        "javax.json;JsonWriter;false;writeObject;;;Argument[0];Argument[-1];taint",
-        "javax.json;JsonWriterFactory;false;createWriter;;;Argument[-1];Argument[0];taint"
-      ]
+      ["javax", "jakarta"] + ".json;" +
+        [
+          "Json;false;createArrayBuilder;(JsonArray);;Argument[0];ReturnValue;taint",
+          "Json;false;createArrayBuilder;(Collection);;Element of Argument[0];ReturnValue;taint",
+          "Json;false;createDiff;;;Argument[0..1];ReturnValue;taint",
+          "Json;false;createMergeDiff;;;Argument[0..1];ReturnValue;taint",
```

Comment:
Not completely sure whether argument 0 (_source_) is actually affecting taint. Did not have a look at the implementation or the RFC, but if I understand it correctly _source_ can only affect whether JSON values are added or removed, but not which values they will have.
E.g.:
- _source_ has `"a": 1`, _target_ does not have it → `a` is removed
- _source_ does not have `a`, _target_ has `"a": 1` → `"a": 1` is added

In both cases _source_ does not taint the patch.
But might be good to verify this, also for JSON arrays.

Reply:
I figured the diff must represent the removal of the key somehow, and `<script>alert('Oh no')</script>` is a valid key, so we might be interested in this as a taint propagator